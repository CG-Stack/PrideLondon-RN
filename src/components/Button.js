// @flow
import React from "react";
import { StyleSheet, Linking } from "react-native";
import Touchable from "./Touchable";
import Text from "./Text";
import {
  eventDetailsBuyButtonColor,
  eventDetailsBuyButtonTextColor,
  buttonPrimaryBackground,
  buttonPrimaryTextColor,
  greyishBrownColor,
  disabledButtonBgColor
} from "../constants/colors";

type ButtonType = "primary";

type Props = {
  text: string,
  url?: string,
  type?: ButtonType,
  onPress?: () => void,
  disabled?: boolean
};

export const handlePress = (onPress: ?() => void, url: ?string) => {
  if (onPress) {
    onPress();
  }

  if (url && url.length > 0) {
    Linking.openURL(url);
  }
};

const Button = ({ text, url, onPress, type = "primary", disabled }: Props) => (
  <Touchable
    style={[styles.button, styles[type], disabled ? styles.disabledButton : {}]}
    onPress={() => handlePress(onPress, url)}
    disabled={disabled}
  >
    <Text
      type="h2"
      style={[
        styles.text,
        styles[`${type}Text`],
        disabled ? styles.disabledText : {}
      ]}
    >
      {text}
    </Text>
  </Touchable>
);
Button.defaultProps = {
  url: "",
  type: "primary",
  onPress: () => {},
  disabled: false
};

const styles = StyleSheet.create({
  button: {
    backgroundColor: eventDetailsBuyButtonColor,
    borderRadius: 4,
    alignItems: "center",
    justifyContent: "center"
  },
  // eslint-disable-next-line react-native/no-unused-styles
  primary: {
    backgroundColor: buttonPrimaryBackground
  },
  // eslint-disable-next-line react-native/no-unused-styles
  primaryText: {
    color: buttonPrimaryTextColor
  },
  text: {
    color: eventDetailsBuyButtonTextColor
  },
  disabledButton: {
    backgroundColor: disabledButtonBgColor
  },
  disabledText: {
    color: greyishBrownColor
  }
});

export default Button;
