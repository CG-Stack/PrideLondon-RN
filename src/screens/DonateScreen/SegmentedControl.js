// @flow
import React from "react";
import { StyleSheet, View } from "react-native";
import Text from "../../components/Text";
import Touchable from "../../components/Touchable";
import {
  inputFieldBorderColor,
  eucalyptusGreenColor
} from "../../constants/colors";

type Props = {|
  onValueChange: (value: string) => void,
  selectedIndex: ?number,
  values: string[]
|};

class SegmentedControl extends React.PureComponent<Props> {
  render() {
    const { onValueChange, selectedIndex, values } = this.props;
    return (
      <View style={styles.container}>
        {values.map((value, index) => (
          <Touchable
            key={value}
            accessibilityComponentType={
              index === selectedIndex
                ? "radiobutton_checked"
                : "radiobutton_unchecked"
            }
            accessibilityTraits={
              index === selectedIndex ? ["button", "selected"] : ["button"]
            }
            onPress={() => onValueChange(value)}
            style={[
              styles.button,
              index === selectedIndex && styles.buttonSelected,
              index === 0 && styles.buttonBorderStart,
              index === values.length - 1 && styles.buttonBorderEnd,
              index - 1 !== selectedIndex && styles.borderButtonLeft
            ]}
          >
            <Text
              type="h2"
              color="lightNavyBlueColor"
              style={styles.buttonText}
            >
              {value}
            </Text>
          </Touchable>
        ))}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flexDirection: "row"
  },
  button: {
    flexGrow: 1,
    borderColor: inputFieldBorderColor,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    height: 50,
    alignItems: "center",
    justifyContent: "center"
  },
  buttonSelected: {
    borderColor: eucalyptusGreenColor,
    backgroundColor: eucalyptusGreenColor
  },
  buttonBorderStart: {
    borderLeftWidth: 1,
    borderTopLeftRadius: 4,
    borderBottomLeftRadius: 4
  },
  buttonBorderEnd: {
    borderRightWidth: 1,
    borderTopRightRadius: 4,
    borderBottomRightRadius: 4
  },
  borderButtonLeft: {
    borderLeftWidth: 1
  },
  buttonText: {
    marginTop: 2
  }
});

export default SegmentedControl;
