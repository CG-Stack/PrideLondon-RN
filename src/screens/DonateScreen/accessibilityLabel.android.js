// @flow

const getAccessibiltyLabel = (
  label: string,
  placeholder: ?string,
  value: ?string,
  focused: boolean
) => {
  const labelParts = [];
  if (focused) {
    labelParts.push("editing");
  }

  if (value) {
    labelParts.push(value);
  }

  labelParts.push("edit box");
  labelParts.push(label);

  const descriptionParts = [];
  if (!focused) {
    descriptionParts.push("double tap to enter text");
  } else {
    descriptionParts.push("double tap and hold to long press");
  }

  return [labelParts.join(", "), descriptionParts.join(", ")].join("; ");
};

export default getAccessibiltyLabel;
