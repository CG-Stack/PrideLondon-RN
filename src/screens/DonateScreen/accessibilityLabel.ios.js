// @flow

const getAccessibiltyLabel = (
  label: string,
  placeholder: ?string,
  value: ?string,
  focused: boolean
) => {
  const labelParts = [];
  const descriptionParts = [];

  labelParts.push(label);

  if (focused) {
    labelParts.push("text field");
    labelParts.push("is editing");

    if (value) {
      labelParts.push(value);
    } else if (placeholder) {
      labelParts.push(placeholder);
    }

    labelParts.push("insertion point at START/END");
  } else {
    if (value) {
      labelParts.push(value);
    } else if (placeholder) {
      labelParts.push(placeholder);
    }

    labelParts.push("text field");

    descriptionParts.push("double tap to edit");
  }

  return [labelParts.join(", "), descriptionParts.join(", ")].join("; ");
};

export default getAccessibiltyLabel;
