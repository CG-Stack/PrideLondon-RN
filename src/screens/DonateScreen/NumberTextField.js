// @flow
import React from "react";
import { Platform, StyleSheet, TextInput } from "react-native";
import {
  lightNavyBlueColor,
  transparent,
  mediumGreyColor,
  eucalyptusGreenColor
} from "../../constants/colors";
import { scaleFont, scaleWithFont } from "../../components/Text";

type Props = {|
  placeholder: string,
  onFocus: () => void,
  onChangeText: string => void,
  returnKeyLabel?: string,
  returnKeyType?: "done" | "go" | "next" | "search" | "send",
  onSubmitEditing?: () => void,
  value: ?string
|};

type State = {|
  focused: boolean
|};

class DonateScreen extends React.PureComponent<Props, State> {
  state = { focused: false };

  onFocus = () => {
    this.setState({ focused: true });
    this.props.onFocus();
  };

  onBlur = () => {
    this.setState({ focused: false });
  };

  render() {
    const { placeholder, onFocus, ...otherProps } = this.props;
    const { focused } = this.state;
    return (
      <TextInput
        keyboardType="numeric"
        onBlur={this.onBlur}
        onFocus={this.onFocus}
        placeholder={focused ? placeholder : null}
        placeholderTextColor={mediumGreyColor}
        style={[styles.base, focused && styles.focused]}
        underlineColorAndroid={transparent}
        {...otherProps}
      />
    );
  }
}

const styles = StyleSheet.create({
  base: {
    borderColor: mediumGreyColor,
    borderWidth: 1,
    borderRadius: 4,
    height: scaleWithFont("h2", 50),
    paddingHorizontal: 14,
    fontFamily: "Poppins-SemiBold",
    fontSize: scaleFont("h2", 18),
    color: lightNavyBlueColor,
    ...Platform.select({
      android: {
        lineHeight: scaleFont("h2", 18),
        includeFontPadding: false,
        textAlignVertical: "bottom"
      }
    })
  },
  focused: {
    borderColor: eucalyptusGreenColor
  }
});

export default DonateScreen;
