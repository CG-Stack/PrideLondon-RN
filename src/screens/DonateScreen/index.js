// @flow
import React from "react";
import {
  KeyboardAvoidingView,
  Linking,
  ScrollView,
  StyleSheet,
  View
} from "react-native";
import type { NavigationScreenProp, NavigationState } from "react-navigation";
import chevronLeftWhite from "../../../assets/images/chevron-left-white.png";
import donateHeader from "../../../assets/images/donateHeader.png";
import Button from "../../components/ButtonPrimary";
import Header from "../../components/Header";
import ImageHeader from "../../components/ImageHeader";
import IconButton from "../../components/IconButton";
import Text from "../../components/Text";
import ContentPadding from "../../components/ContentPadding";
import { lightNavyBlueColor, whiteColor } from "../../constants/colors";
import text from "../../constants/text";
import NumberTextField from "./NumberTextField";
import SegmentedControl from "./SegmentedControl";

type Props = {|
  navigation: NavigationScreenProp<NavigationState>
|};

type State = {|
  selectedAmount: ?string,
  otherAmount: ?string
|};

const selectableAmounts = ["£5", "£10", "£20"];

class DonateScreen extends React.PureComponent<Props, State> {
  static navigationOptions = {
    tabBarVisible: false
  };

  state = {
    selectedAmount: null,
    otherAmount: null
  };

  onAmountSelected = (selectedAmount: string) => {
    this.setState({ selectedAmount, otherAmount: null });
  };

  onOtherAmountFocus = () => {
    this.setState({ selectedAmount: null });
  };

  onOtherAmountChange = (otherAmount: string) => {
    this.setState({ otherAmount });
  };

  onDonatePress = () => {
    const amount = this.state.amount || "0";
    Linking.openURL(
      `https://donate.prideinlondon.org/?amount=${encodeURIComponent(amount)}`
    );
  };

  render() {
    const { selectedAmount, otherAmount } = this.state;

    return (
      <View style={styles.container}>
        <Header backgroundColor={lightNavyBlueColor}>
          <ContentPadding style={styles.headerContent}>
            <IconButton
              accessibilityLabel="Back"
              onPress={() => {
                this.props.navigation.goBack(null);
              }}
              source={chevronLeftWhite}
              testID="back"
            />
            <Text type="h2" style={styles.headerTitle}>
              {text.donateTitle}
            </Text>
            <View style={styles.phantomIcon} />
          </ContentPadding>
        </Header>
        <ScrollView>
          <KeyboardAvoidingView
            behavior="position"
            enabled
            style={styles.scrollContainer}
          >
            <ImageHeader image={donateHeader} title={text.donateHeader} />
            <ContentPadding>
              <Text
                type="h2"
                color="lightNavyBlueColor"
                style={styles.introHeadingSpacing}
              >
                {text.donateIntroductionHeading}
              </Text>
              <Text>{text.donateIntroduction}</Text>
              <Text
                type="h3"
                color="lightNavyBlueColor"
                style={styles.amountSelectionlabelSpacing}
              >
                {text.donateAmountSelectionLabel}
              </Text>
              <SegmentedControl
                onValueChange={this.onAmountSelected}
                selectedIndex={selectableAmounts.indexOf(selectedAmount)}
                values={selectableAmounts}
              />
              <Text
                type="h4"
                color="lightNavyBlueColor"
                style={styles.otherAmountlabelSpacing}
              >
                {text.donateOtherAmountLabel}
              </Text>
              <NumberTextField
                placeholder={(0).toFixed(2)}
                onFocus={this.onOtherAmountFocus}
                onChangeText={this.onOtherAmountChange}
                value={otherAmount}
              />
              <Text type="small" style={styles.minimumAmountSpacing}>
                {text.donateMinimumAmount}
              </Text>
              <View style={styles.ctaSpacing}>
                <Button
                  disabled={!selectedAmount && !otherAmount}
                  onPress={this.onDonatePress}
                >
                  {text.donateButtonText}
                </Button>
              </View>
            </ContentPadding>
          </KeyboardAvoidingView>
        </ScrollView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: whiteColor
  },
  headerContent: {
    width: "100%",
    maxWidth: 440,
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
    alignSelf: "center"
  },
  headerTitle: {
    color: whiteColor
  },
  phantomIcon: {
    width: 48,
    height: 48
  },
  scrollContainer: {
    width: "100%",
    maxWidth: 440,
    alignSelf: "center"
  },
  introHeadingSpacing: {
    marginTop: 12
  },
  amountSelectionlabelSpacing: {
    marginTop: 20,
    marginBottom: 16
  },
  otherAmountlabelSpacing: {
    marginTop: 20,
    marginBottom: 8
  },
  minimumAmountSpacing: {
    marginTop: 10
  },
  ctaSpacing: {
    marginTop: 16,
    paddingVertical: 12
  }
});

export default DonateScreen;
