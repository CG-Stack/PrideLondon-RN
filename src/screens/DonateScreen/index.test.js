// @flow
import React from "react";
import { Linking } from "react-native";
import { shallow } from "enzyme";
import DonateScreen from ".";
import Button from "../../components/ButtonPrimary";
import NumberTextField from "./NumberTextField";
import SegmentedControl from "./SegmentedControl";

let openURLSpy;
beforeEach(() => {
  openURLSpy = jest
    .spyOn(Linking, "openURL")
    .mockImplementation(() => Promise.resolve());
});

it("renders correctly", () => {
  const output = shallow(<DonateScreen navigation={null} />);
  expect(output).toMatchSnapshot();
});

it("sets selected and clears other amound when using the amount selector", () => {
  const output = shallow(<DonateScreen navigation={null} />);
  const amountSelector = output.find(SegmentedControl);

  output.setState({ otherAmount: "20" });
  amountSelector.prop("onSelectedIndexChange")(1);
  expect(output.state()).toEqual({
    otherAmount: null,
    selectedAmount: 1
  });
});

it("clears selected amount when focusing the other amount input", () => {
  const output = shallow(<DonateScreen navigation={null} />);
  const otherAmount = output.find(NumberTextField);

  output.setState({ selectedAmount: 1 });
  otherAmount.prop("onFocus")();
  expect(output.state()).toEqual({
    otherAmount: null,
    selectedAmount: null
  });
});

it("sets other amound when using the other amount input", () => {
  const output = shallow(<DonateScreen navigation={null} />);
  const otherAmount = output.find(NumberTextField);

  otherAmount.prop("onChangeText")("4");
  expect(output.state()).toEqual({
    otherAmount: "4",
    selectedAmount: null
  });
});

it("opens donation website with default amount when pressing donate button", () => {
  const output = shallow(<DonateScreen navigation={null} />);
  output.find(Button).simulate("press");
  expect(openURLSpy).toHaveBeenCalledWith(
    "https://donate.prideinlondon.org/?amount=0"
  );
});

it("opens donation website with selected amount when pressing donate button", () => {
  const output = shallow(<DonateScreen navigation={null} />);
  output.setState({ selectedAmount: 1 });
  output.find(Button).simulate("press");
  expect(openURLSpy).toHaveBeenCalledWith(
    "https://donate.prideinlondon.org/?amount=10"
  );
});

it("opens donation website with other amount when pressing donate button", () => {
  const output = shallow(<DonateScreen navigation={null} />);
  output.setState({ otherAmount: "0&hack=1" });
  output.find(Button).simulate("press");
  expect(openURLSpy).toHaveBeenCalledWith(
    "https://donate.prideinlondon.org/?amount=0%26hack%3D1"
  );
});

it("navigates back when user presses back button in toolbar", () => {
  const navigation = { goBack: jest.fn() };
  const output = shallow(<DonateScreen navigation={navigation} />);
  output.find({ testID: "back" }).simulate("press");
  expect(navigation.goBack).toHaveBeenCalledWith(null);
});

afterEach(() => {
  openURLSpy.mockRestore();
});
