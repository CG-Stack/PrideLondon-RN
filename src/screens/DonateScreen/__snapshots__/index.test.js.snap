// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <Header
    backgroundColor="#2d2f7f"
  >
    <ContentPadding
      padding={Object {}}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "maxWidth": 440,
          "width": "100%",
        }
      }
    >
      <IconButton
        accessibilityLabel="Back"
        onPress={[Function]}
        source={1}
        testID="back"
      />
      <Text
        color="blackColor"
        markdown={false}
        markdownStyle={Object {}}
        style={
          Object {
            "color": "#ffffff",
          }
        }
        type="h2"
      >
        Donate
      </Text>
      <Component
        style={
          Object {
            "height": 48,
            "width": 48,
          }
        }
      />
    </ContentPadding>
  </Header>
  <KeyboardAvoidingView
    behavior="padding"
    enabled={true}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ScrollView>
      <Component
        style={
          Object {
            "alignSelf": "center",
            "maxWidth": 440,
            "width": "100%",
          }
        }
      >
        <ImageHeader
          image={1}
          title={
            Array [
              "\\"It's the first time I feel",
              "accepted in a big group.\\"",
            ]
          }
        />
        <ContentPadding
          padding={Object {}}
          style={Object {}}
        >
          <Text
            color="lightNavyBlueColor"
            markdown={false}
            markdownStyle={Object {}}
            style={
              Object {
                "marginTop": 12,
              }
            }
            type="h2"
          >
            Be part of the movement that drives diversity
          </Text>
          <Text
            color="blackColor"
            markdown={false}
            markdownStyle={Object {}}
            type="text"
          >
            The need for Pride is as strong as ever. A huge amount of progress has been made for LGBT+ people but challenges and stigma are still rife. Whether you take part to celebrate, party or protest. Pride in London is your platform to make change. Your donations keep us marching.
          </Text>
          <Text
            color="lightNavyBlueColor"
            markdown={false}
            markdownStyle={Object {}}
            style={
              Object {
                "marginBottom": 16,
                "marginTop": 20,
              }
            }
            type="h3"
          >
            Make a single donation
          </Text>
          <SegmentedControl
            onValueChange={[Function]}
            selectedValue={null}
          >
            <Component
              key="5"
              label="£5"
              value={5}
            />
            <Component
              key="10"
              label="£10"
              value={10}
            />
            <Component
              key="20"
              label="£20"
              value={20}
            />
          </SegmentedControl>
          <Text
            color="lightNavyBlueColor"
            markdown={false}
            markdownStyle={Object {}}
            style={
              Object {
                "marginBottom": 8,
                "marginTop": 20,
              }
            }
            type="h4"
          >
            Other amount (£)
          </Text>
          <DonateScreen
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="0.00"
            value={null}
          />
          <Text
            color="blackColor"
            markdown={false}
            markdownStyle={Object {}}
            style={
              Object {
                "marginTop": 10,
              }
            }
            type="small"
          >
            The minimum donation amount is £2.
          </Text>
          <Component
            style={
              Object {
                "marginTop": 16,
                "paddingVertical": 12,
              }
            }
          >
            <Button
              disabled={true}
              onPress={[Function]}
            >
              Donate
            </Button>
          </Component>
        </ContentPadding>
      </Component>
    </ScrollView>
  </KeyboardAvoidingView>
</Component>
`;
